proposals/pending-state.md

```diff
@@ -0,0 +1,163 @@
+# pending-task-protocol
+
+An open protocol for interoperable asynchronous Web Components.
+
+Author: Justin Fagnani
+
+Status: Draft
+
+Last update: 2021-066-20
+
+# Background
+
+There are a number of scenarios where a web component might depend on or perform some asynchronous task. Components may lazy-load parts of their implementation or content, perform I/O in response to user input, manage long-running computations, etc.
+
+It's often desirable to communicate the state of async tasks up the component tree to parent components so that they can display user affordances indicating whether the UI or content is pending some async operation.
+
+Frameworks have the ability to invent custom APIs and patterns to handle this kind of cross-component communication. To enable the same with web components in a way that's interoperable between components from diffrent sources, and possibly implemented with different libraries, we need a protocol that components can implement without reliance on a common implementation.
+
+# Goals
+
+* Allow components to communicate that they have pending asynchonous tasks to ancestors in the DOM tree.
+* Allow components to know if there are pending asynchronous tasks in descendants in the DOM tree.
+* Allow components to intercept, modify, and block pending task notifications.
+* Give guidance on the type of asynchronous work that should be notified.
+* Allow for some labelling or differentiation between types of asynchronous work.
+
+# Details
+
+## Asynchronous Task States
+
+There are four main states that an asynchronous task can be in:
+
+* Not-started
+* Started
+* Completed
+* Failed
+
+This protocol represents three of the states (Started, Completed, and Failed) with a promise-carrying DOM event.
+
+## The `pending-task` event
+
+Components with pending tasks indicate so by firing a composed, bubbling, `pending-task` Event, with a `promise` property:
+
+TypeScript interface:
+```ts
+interface PendingTaskEvent extends Event {
+  complete: Promise<void>;
```

Comment:
```suggestion
  complete: Promise<void> | Promise<void>[];
```
This would make it possible to bundle multiple pending tasks in a single event, so that we don't show the spinner when one task has completed but there are other pending tasks still pending. Then on the indicator's end it would be something like `Array.isArray(e.complete) ? await Promise.all(e.complete) : await e.complete;`

Alternatively, I suppose you could do `this.dispatchEvent(new PendingTaskEvent(Promise.all(this.pendingTaskOne, this.pendingTaskTwo)));` which is perhaps simpler 🤔 I don't have a strong preference but I thought I'd bring up the question of handling multiple pending tasks

Reply:
As we discussed in chat: I don't think that we want the event to vary on the type of `.complete`. If dispatchers have multiple pending tasks they can either send multiple events or call `Promise.all()` themselves.